### ISSUE TITLE
getTypeDeclaration() thread safety

### ISSUE THREAD

**Issue by Egor18** [INRIA/spoon#2180]
As far as I understand, spoon model should be thread safe for reading.
However, in some rare cases I face problems when I call `getTypeDeclaration()` from different threads.

The error looks like this:
```
java.lang.ClassCastException: java.util.HashMap$Node cannot be cast to java.util.HashMap$TreeNode
	at java.util.HashMap$TreeNode.moveRootToFront(HashMap.java:1832)
	at java.util.HashMap$TreeNode.treeify(HashMap.java:1949)
	at java.util.HashMap$TreeNode.split(HashMap.java:2175)
	at java.util.HashMap.resize(HashMap.java:714)
	at java.util.HashMap.putVal(HashMap.java:663)
	at java.util.HashMap.put(HashMap.java:612)
	at spoon.reflect.factory.TypeFactory.get(TypeFactory.java:555)
	at spoon.support.reflect.reference.CtTypeReferenceImpl.getTypeDeclaration(CtTypeReferenceImpl.java:186)
        ...
```
I believe that the possible solution is to use `ConcurrentHashMap` instead of `HashMap` for `shadowCache` in `TypeFactory`.
Unfortunately it's hard to reproduce this error, because it occurs only sometimes.

So, what do you think about it?
Thanks.

**Comment by surli** [comment]
Maybe we could use something like https://github.com/ThomasKrieger/concurrent-junit to reproduce the error in a test? WDYT?

**Comment by Egor18** [comment]
It could be possible, to write such kind of test, but I'm not sure.

The point is that you need 2 different threads, which are calling getTypeDeclaration from the same type reference at the same time, so they both enter here:
https://github.com/INRIA/spoon/blob/10e983b20f7bbf793e87d965f2e1730f48f35bc8/src/main/java/spoon/reflect/factory/TypeFactory.java#L555

BTW, does it seem like a bug to you, or `getTypeDeclaration()` is not supposed to be thread safe?

**Comment by monperrus** [comment]
Spoon has never committed to provide concurrency safety.

That said, it would be great to do so on some methods.

**Comment by Egor18** [comment]
I can make the PR to change the type to `ConcurrentHashMap`. WDYT?

**Comment by monperrus** [comment]
good idea!

**Comment by pvojtechovsky** [comment]
fixed by #2198
